lib/bolt/plugin/puppetdb.rb

```diff
@@ -15,9 +24,66 @@ def hooks
         ['lookup_targets']
       end
 
+      def warn_missing_fact(certname, fact)
+        @logger.warn("Could not find fact #{fact} for node #{certname}")
+      end
+
+      def fact_path(raw_fact)
+        fact = raw_fact.split(".")
+        raise FactLookupError, raw_fact unless fact[0] == 'facts'
+        fact.drop(1)
+      end
+
       def lookup_targets(opts)
-        nodes = @puppetdb_client.query_certnames(opts['query'])
-        nodes.map { |certname| { 'uri' => certname } }
+        targets = @puppetdb_client.query_certnames(opts['query'])
+        facts = []
+
+        target_opts = opts.select { |k, _| TARGET_OPTS.include?(k) }
+        Bolt::Util.walk_vals(target_opts) do |value|
+          # This is done in parts instead of in place so that we only need to
+          # make one puppetDB query
+          if value.is_a?(String)
+            facts << fact_path(value)
+          end
+          value
+        end
+
+        facts.uniq!
+        # Returns {'mycertname' => [{'path' => ['nested', 'fact'], 'value' => val'}], ... }
+        fact_values = facts.empty? ? {} : @puppetdb_client.fact_values(targets, facts)
+
+        targets.map do |certname|
+          target_data = fact_values[certname]
+
+          target = {}
+          target_opts.each do |k, v|
```

Comment:
I may be missing but I think we just want to pass `target_opts` directly to `resolve_facts` and let it `walk_vals` rather than looping over them.